This script contain the following
1. New notebook dataframe, that include newly derived columns from the previous Exercise.
2. Repeated process for the entire dataframe
3. Analyze result and how results for the entire dataframe differ from those of the subset.
4. Creating loyalty flag for existing customers using the transform() and loc() functions.
5. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer)
6. Created spending flag for each user based on the average price across all their orders.
7. Created order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. 
8. Exported dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [5]:
# import libraries
import pandas as pd
import numpy as np
import os

In [6]:
path = r'C:\Users\Tim\Desktop\Data Analytics Intro\Python Fundamentals for Data Analysts 4\Instacart Basket Analysis'

In [7]:
path

'C:\\Users\\Tim\\Desktop\\Data Analytics Intro\\Python Fundamentals for Data Analysts 4\\Instacart Basket Analysis'

In [8]:
# import merge dataframe
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_derived.pkl'))

# 2. Repeat process for the entire dataframe

In [10]:
ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.457687
2     17.277920
3     17.179756
4     17.811403
5     15.213779
6     16.439806
7     17.225773
8     15.340520
9     15.895474
10    20.197148
11    16.170828
12    15.887622
13    16.583304
14    16.757377
15    16.165037
16    17.663250
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

The results from the complete dataframe include EVERY 'department_id' value in contrast to the results from the subset. All of the department ids in the subset now have different mean values. While some have decreased, others have increased. For instance, department id 10's mean value in the subset was '18.681852', but in the new dataframe, it is '20.197148'.


# 4. Creating loyalty flag for existing customers using the transform() and loc() functions.

In [11]:
# Aggregating Data with transform()
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\Tim\AppData\Local\Temp\ipykernel_25564\598408120.py:2: FutureWarning: The provided callable <function max at 0x000002387888B380> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [12]:
ords_prods_merge.head(20)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    3139998      138            28                   6                 11   
1    1977647      138            30                   6                 17   
2     389851      709             2                   0                 21   
3     652770      764             1                   3                 13   
4    1813452      764             3                   4                 17   
5    1701441      777            16                   1                  7   
6    1871483      825             3                   2                 14   
7    1290456      910            12                   3                 10   
8     369558     1052            10                   1                 20   
9     589712     1052            15                   1                 12   
10   1798115     1379             1                   5                 11   
11    497193     1480            36                   3                 13   
12   1727826     1480            38                   4                 13   
13   2694698     1480            39                   2                 12   
14   3171753     1494             6                   3                 10   
15    186761     1494             7                   2                  8   
16    317121     1494             9                   5                  8   
17    200390     1540             4                   2                  0   
18    374436     1540             5                   1                 21   
19   2920553     1540             6                   1                  0   

    days_since_prior_order  add_to_cart_order  reordered _merge  \
0                      3.0                  5          0   both   
1                     20.0                  1          1   both   
2                      6.0                 20          0   both   
3                      NaN                 10          0   both   
4                      9.0                 11          1   both   
5                     26.0                  7          0   both   
6                     30.0                  2          0   both   
7                     30.0                  1          0   both   
8                     19.0                  1          0   both   
9                     15.0         

In [13]:
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [14]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [16]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [17]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15891077
Loyal customer      10293737
New customer         6249398
Name: count, dtype: int64

In [18]:
# View updated data frame
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     NaN                 10          0   both   
4                     9.0                 11          1   both   

     price_range_loc     busiest_day    busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regularly busy           Most orders   
1  Mid-range product  Regularly busy  Regularly busy        Average orders   
2  Mid-range product     Busiest day    Busiest days        Average orders   
3  Mid-range product  Regularly busy    Slowest days           Most orders   
4  Mid-range product      Least busy    Slowest days        Average orders   

   max_order      loyalty_flag  
0         32  Regular customer  
1         32  Regular customer  
2          5      New customer  
3          3      New customer  
4          3      New customer

# 6. Created spending flag for each user based on the average price across all their orders using the following criteria.

In [19]:
# Use groupby and agg() to determine basic statisitcs of spending habits by loyalty group
ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.388747  1.0  99999.0
New customer      13.294370  1.0  99999.0
Regular customer  12.496203  1.0  99999.0

The average price of a product purchased by a new customer is about 13 more than that of a loyal and regualar customer.

6. Create a spending flag for each user based on the average price across all their orders using the following criteria: If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.” If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [20]:
# Use transform() to determine the mean price of products purchased by each user
ords_prods_merge['mean_product_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\Tim\AppData\Local\Temp\ipykernel_25564\4025210921.py:2: FutureWarning: The provided callable <function mean at 0x000002387888BD80> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['mean_product_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [21]:
# View relevant columns 
ords_prods_merge[['user_id', 'mean_product_price']].head(20)

user_id  mean_product_price
0       138            6.935811
1       138            6.935811
2       709            7.930208
3       764            4.972414
4       764            4.972414
5       777            6.935398
6       825            5.957576
7       910            6.680000
8      1052            7.162500
9      1052            7.162500
10     1379            5.813636
11     1480            5.139394
12     1480            5.139394
13     1480            5.139394
14     1494            6.803797
15     1494            6.803797
16     1494            6.803797
17     1540            6.607122
18     1540            6.607122
19     1540            6.607122

# Create spending_flag column with given criteria using loc() function

In [22]:
ords_prods_merge.loc[ords_prods_merge['mean_product_price'] >= 10, 'spending_flag'] = 'High spender'

In [23]:
ords_prods_merge.loc[ords_prods_merge['mean_product_price'] < 10, 'spending_flag'] = 'Low spender'

In [24]:
ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     31798751
High spender      635461
Name: count, dtype: int64

In [25]:
# View relevant columns of updated data frame
ords_prods_merge[['user_id', 'mean_product_price', 'spending_flag']].head(20)

user_id  mean_product_price spending_flag
0       138            6.935811   Low spender
1       138            6.935811   Low spender
2       709            7.930208   Low spender
3       764            4.972414   Low spender
4       764            4.972414   Low spender
5       777            6.935398   Low spender
6       825            5.957576   Low spender
7       910            6.680000   Low spender
8      1052            7.162500   Low spender
9      1052            7.162500   Low spender
10     1379            5.813636   Low spender
11     1480            5.139394   Low spender
12     1480            5.139394   Low spender
13     1480            5.139394   Low spender
14     1494            6.803797   Low spender
15     1494            6.803797   Low spender
16     1494            6.803797   Low spender
17     1540            6.607122   Low spender
18     1540            6.607122   Low spender
19     1540            6.607122   Low spender

# Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column.

In [26]:
# Use groupby and transform to determine median number of days between orders for each user
ords_prods_merge['median_days_between_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

C:\Users\Tim\AppData\Local\Temp\ipykernel_25564\2923105074.py:2: FutureWarning: The provided callable <function median at 0x00000238789D39C0> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['median_days_between_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [27]:
ords_prods_merge[['user_id', 'days_since_prior_order', 'median_days_between_orders']].head(20)

user_id  days_since_prior_order  median_days_between_orders
0       138                     3.0                         8.0
1       138                    20.0                         8.0
2       709                     6.0                         8.0
3       764                     NaN                         9.0
4       764                     9.0                         9.0
5       777                    26.0                        11.0
6       825                    30.0                        20.0
7       910                    30.0                         6.0
8      1052                    19.0                        10.0
9      1052                    15.0                        10.0
10     1379                     NaN                        20.0
11     1480                     7.0                         7.0
12     1480                     7.0                         7.0
13     1480                    12.0                         7.0
14     1494                     2.0                         7.0
15     1494                     6.0                         7.0
16     1494                    10.0                         7.0
17     1540                     7.0                         7.0
18     1540                     6.0                         7.0
19     1540                    14.0                         7.0

# # Create order frequency flag based on given criteria using loc function

In [28]:
ords_prods_merge.loc[ords_prods_merge['median_days_between_orders'] > 20, 'order_frequency'] = 'Non-frequent customer'

In [29]:
ords_prods_merge.loc[(ords_prods_merge['median_days_between_orders'] <= 20) & (ords_prods_merge['median_days_between_orders'] > 10), 'order_frequency'] = 'Regular customer'

In [30]:
ords_prods_merge.loc[ords_prods_merge['median_days_between_orders'] <= 10, 'order_frequency'] = 'Frequent customer'

In [31]:
# Check value counts of new order_frequency column
ords_prods_merge['order_frequency'].value_counts(dropna = False)

order_frequency
Frequent customer        21577409
Regular customer          7217134
Non-frequent customer     3639669
Name: count, dtype: int64

In [32]:
# View relevant columns of updated data frame
ords_prods_merge[['user_id', 'days_since_prior_order', 'median_days_between_orders', 'order_frequency']].head(25)

user_id  days_since_prior_order  median_days_between_orders  \
0       138                     3.0                         8.0   
1       138                    20.0                         8.0   
2       709                     6.0                         8.0   
3       764                     NaN                         9.0   
4       764                     9.0                         9.0   
5       777                    26.0                        11.0   
6       825                    30.0                        20.0   
7       910                    30.0                         6.0   
8      1052                    19.0                        10.0   
9      1052                    15.0                        10.0   
10     1379                     NaN                        20.0   
11     1480                     7.0                         7.0   
12     1480                     7.0                         7.0   
13     1480                    12.0                         7.0   
14     1494                     2.0                         7.0   
15     1494                     6.0                         7.0   
16     1494                    10.0                         7.0   
17     1540                     7.0                         7.0   
18     1540                     6.0                         7.0   
19     1540                    14.0                         7.0   
20     1540                     8.0                         7.0   
21     1540                     6.0                         7.0   
22     1540                    28.0                         7.0   
23     1540                     7.0                         7.0   
24     1540                     7.0                         7.0   

      order_frequency  
0   Frequent customer  
1   Frequent customer  
2   Frequent customer  
3   Frequent customer  
4   Frequent customer  
5    Regular customer  
6    Regular customer  
7   Frequent customer  
8   Frequent customer  
9   Frequent customer  
10   Regular customer  
11  Frequent customer  
12  Frequent customer  
13  Frequent customer  
14  Frequent customer  
15  Frequent customer  
16  Frequent customer  
17  Frequent customer  
18  Frequent customer  
19  Frequent customer  
20  Frequent customer  
21  Frequent customer  
22  Frequent customer  
23  Frequent customer  
24  Frequent customer

# 7. Export as pickle file 

In [34]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merged_grouped.pkl'))